metrics/web/lib/project_groups/presentation/state/project_groups_notifier.dart

```diff
@@ -0,0 +1,329 @@
+    );
+
+    final projectIds = projectGroup?.projectIds ?? [];
+
+    _projectSelectorViewModels = _projectSelectorViewModels
+        .map(
+          (project) => ProjectSelectorViewModel(
+            id: project.id,
+            name: project.name,
+            isChecked: projectIds?.contains(project.id),
+          ),
+        )
+        .toList();
+
+    _activeProjectGroupDialogViewModel = ActiveProjectGroupDialogViewModel(
+      id: projectGroup?.id,
+      name: projectGroup?.name,
+      selectedProjectIds: List<String>.from(projectIds),
+    );
+
+    notifyListeners();
+  }
+
+  /// Change checked status for [ProjectSelectorViewModel] by [projectId].
+  void toggleProjectCheckedStatus({String projectId, bool isChecked}) {
+    final projectIds = _activeProjectGroupDialogViewModel.selectedProjectIds;
+
+    if (isChecked) {
+      projectIds.add(projectId);
+    } else {
+      projectIds.remove(projectId);
+    }
+
+    final projectIndex = _projectSelectorViewModels
+        .indexWhere((project) => project.id == projectId);
+
+    final project = _projectSelectorViewModels[projectIndex];
+
+    _projectSelectorViewModels[projectIndex] = ProjectSelectorViewModel(
+      id: project.id,
+      name: project.name,
+      isChecked: isChecked,
+    );
+
+    _activeProjectGroupDialogViewModel = ActiveProjectGroupDialogViewModel(
+      id: _activeProjectGroupDialogViewModel.id,
+      name: _activeProjectGroupDialogViewModel.name,
+      selectedProjectIds: projectIds,
+    );
+
+    notifyListeners();
+  }
+
+  /// Subscribes to project groups.
+  Future<void> subscribeToProjectGroups() async {
+    final projectGroupsStream = _receiveProjectGroupUpdates();
+    _errorMessage = null;
+    await _projectGroupsSubscription?.cancel();
+    _projectGroupsSubscription = projectGroupsStream.listen(
+      _projectGroupListener,
+      onError: _errorHandler,
+    );
+  }
+
+  /// Unsubscribes from project groups.
+  Future<void> unsubscribeFromProjectGroups() async {
+    await _cancelSubscriptions();
+    notifyListeners();
+  }
+
+  /// Saves project group data into Firestore with the given [projectGroupName],
+  /// [projectIds].
+  ///
+  /// If [projectIds] is null, a new project group is added,
+  /// otherwise existing ones are updated.
+  Future<bool> saveProjectGroup(
+    String projectGroupId,
+    String projectGroupName,
+    List<String> projectIds,
+  ) async {
+    resetFirestoreWriteErrorMessage();
+
+    try {
+      if (projectGroupId == null) {
+        await _addProjectGroupUseCase(
+          AddProjectGroupParam(
+            projectGroupName: projectGroupName,
+            projectIds: projectIds,
+          ),
+        );
+      } else {
+        await _updateProjectGroupUseCase(
+          UpdateProjectGroupParam(
+            projectGroupId,
+            projectGroupName,
+            projectIds,
+          ),
+        );
+      }
+    } catch (e) {
+      _firestoreWriteErrorHandler(e);
+    }
```

Comment:
If we want to catch any errors from the repository, we should do it in the repository itself and then wrap/adapt it to be a metrics specific exception. This allows us to be independent of the repository implementation. See examples of implementation in the `auth` module.